evm/auxiliary/user_account_contract/contract.py

```diff
@@ -0,0 +1,211 @@
+(
+    CALLDATA_V,
+    CALLDATA_R,
+    CALLDATA_S,
+    CALLDATA_NONCE,
+    CALLDATA_GASPRICE,
+    CALLDATA_VALUE,
+    CALLDATA_MIN_BLOCK,
+    CALLDATA_MAX_BLOCK,
+    CALLDATA_DESTINATION,
+    CALLDATA_DATA,
+) = (CALLDATA_SIGNATURE + i * 32 for i in range(10))
+
+# memory locations
+MEMORY_CALLDATA = 32
+MEMORY_NONCE = MEMORY_CALLDATA + CALLDATA_NONCE
+MEMORY_DATA = MEMORY_CALLDATA + CALLDATA_DATA
+MEMORY_SIGHASH = 0
+MEMORY_SENDER = 0
+MEMORY_RETURNCODE = 0
+
+# storage locations
+STORAGE_NONCE = 0
+
+
+def generate_account_lll_code(validation_code_address):
+    return [
+        'seq',
+
+        # handle get_nonce if called
+        ['if',
+            ['eq',
+                ['calldataload', CALLDATA_FUNCTION_ID],
+                NONCE_GETTER_ID],
+            [
+                'seq',
+                ['mstore', 0, ['sload', STORAGE_NONCE]],
+                ['return', 0, 32]
+            ]],
+
+        # copy full call data to memory
+        ['calldatacopy', MEMORY_CALLDATA, 0, ['calldatasize']],
+
+        # handle get_validation_code if called
+        ['if',
+            ['eq',
+                ['calldataload', CALLDATA_FUNCTION_ID],
+                VALIDATION_CODE_GETTER_ID,
+                # return validation code address
+                ['mstore', 0, validation_code_address],
+                ['return', 12, 20]]],
+
+        # no function is called, so we should be first in the call stack
+        ['assert', ['eq', ['caller'], ENTRY_POINT_INT]],
+
+        # check if block is in valid range
+        ['assert', ['ge', ['number'], ['calldataload', CALLDATA_MIN_BLOCK]]],
+        ['assert', ['le', ['number'], ['calldataload', CALLDATA_MAX_BLOCK]]],
+
+        # Check for small s to avoid malleability
+        ['assert', ['lt', ['calldataload', CALLDATA_S], MAX_ALLOWED_S + 1]],
+
+        # load tx-sighash to the end of memory
+        ['mstore', ['add', MEMORY_CALLDATA, ['calldatasize']], ['sighash']],
+        # Compute sighash = sha3(nonce ++ gasprice ++ value ++ min-block ++ max-block ++ to ++
+        # data ++ tx-sighash)
+        ['mstore',
+            MEMORY_SIGHASH,
+            ['sha3', MEMORY_NONCE, ['sub', ['calldatasize'], CALLDATA_NONCE - 32]]],
+
+        # Verify signature by calling validation code
+        ['assert',
+            [
+                'call',
+                VALIDATION_CODE_GAS,
+                validation_code_address,
+                0,
+                # input data (spans sighash and signature)
+                MEMORY_SIGHASH,
+                128,
+                # output data (discarded)
+                0,
+                0
+            ]],
+
+        # Verify and increment nonce
+        ['assert', ['eq', ['calldataload', CALLDATA_NONCE], ['sload', STORAGE_NONCE]]],
+        ['sstore', STORAGE_NONCE, ['add', ['calldataload', CALLDATA_NONCE], 1]],
+
+        # Assert that we won't run out of gas from here on
+        ['assert', ['gt', 'gas', GAS_RESERVE + GAS_RESERVE_OFFSET]],
```

Comment:
Is this line still necessary?

Reply:
Depends: If we run out of gas after the PAYGAS call, the transaction can be replayed infinitely and the account would get drained. This line makes sure that this can never happen. If we trust the user to never create transactions with not enough gas, the check isn't necessary. But at least I would feel much safer with it.